File: sample/kalman_filter/extended_kalman_filter.py
"""
File: extended_kalman_filter.py

This script demonstrates the implementation and simulation of an Extended Kalman Filter
(EKF) for a bicycle model with delayed measurements.
The EKF is designed to estimate the state of a vehicle (position and orientation)
using noisy sensor measurements from two landmarks.
The code symbolically defines the system and measurement models,
generates their Jacobians, and deploys them for use in the EKF.
It also provides a simulation environment to test the EKF's performance
under realistic conditions, including process and measurement noise, as well as system delays.

Reference URL:
https://inzkyk.xyz/kalman_filter/extended_kalman_filters/#subsection:11.4.1
"""
import os
import sys
sys.path.append(os.getcwd())

import math
import numpy as np
import matplotlib.pyplot as plt
import sympy
from sympy import symbols
from dataclasses import dataclass

from external_libraries.MCAP_python_control.python_control.kalman_filter import ExtendedKalmanFilter
from external_libraries.MCAP_python_control.python_control.control_deploy import ExpressionDeploy
from python_control.kalman_filter_deploy import KalmanFilterDeploy

from sample.simulation_manager.visualize.simulation_plotter import SimulationPlotter


@dataclass
class Parameters:
    delta_time: float
    wheelbase: float
    landmark_1_x: float
    landmark_1_y: float
    landmark_2_x: float
    landmark_2_y: float


def main():
    # %% bicycle model example
    # state X: [x, y, theta]
    # input U: [v, steering_angle]
    # output Y: [r_p, angle_p]

    delta_time = symbols('delta_time')
    steering_angle = symbols('steering_angle')
    x = symbols('x')
    y = symbols('y')
    v = symbols('v')
    wheelbase = symbols('wheelbase')
    radius = symbols('radius')
    theta = symbols('theta')
    landmark_1_x = symbols('landmark_1_x')
    landmark_1_y = symbols('landmark_1_y')
    landmark_2_x = symbols('landmark_2_x')
    landmark_2_y = symbols('landmark_2_y')

    # define state X, input U
    X = sympy.Matrix([[x], [y], [theta]])
    U = sympy.Matrix([[v], [steering_angle]])

    d = v * delta_time
    beta = (d / wheelbase) * sympy.tan(steering_angle)
    r = wheelbase / sympy.tan(steering_angle)

    fxu = sympy.Matrix([[x - r * sympy.sin(theta) + r * sympy.sin(theta + beta)],
                        [y + r * sympy.cos(theta) - r *
                         sympy.cos(theta + beta)],
                        [theta + beta]])
    fxu_jacobian = fxu.jacobian(X)
    print("fxu_jacobian:\n", fxu_jacobian)

    hx = sympy.Matrix([[sympy.sqrt((landmark_1_x - x) ** 2 + (landmark_1_y - y) ** 2)],
                       [sympy.atan2(landmark_1_y - y,
                                    landmark_1_x - x) - theta],
                       [sympy.sqrt((landmark_2_x - x) ** 2 +
                                   (landmark_2_y - y) ** 2)],
                       [sympy.atan2(landmark_2_y - y, landmark_2_x - x) - theta]])
    hx_jacobian = hx.jacobian(X)
    print("hx_jacobian:\n", hx_jacobian)

    # Save functions to separate files
    ExpressionDeploy.write_state_function_code_from_sympy(fxu, X, U)
    ExpressionDeploy.write_state_function_code_from_sympy(fxu_jacobian, X, U)

    ExpressionDeploy.write_measurement_function_code_from_sympy(hx, X)
    ExpressionDeploy.write_measurement_function_code_from_sympy(
        hx_jacobian, X)

    # %% design EKF

    landmarks = np.array([[-1.0, 10.0], [-1.0, 10.0]])

    Parameters_ekf = Parameters(
        delta_time=0.1, wheelbase=0.5,
        landmark_1_x=landmarks[0, 0], landmark_1_y=landmarks[1, 0],
        landmark_2_x=landmarks[0, 1], landmark_2_y=landmarks[1, 1])

    Number_of_Delay = 5

    Q_ekf = np.diag([1.0, 1.0, 1.0])
    R_ekf = np.diag([1.0, 1.0, 1.0, 1.0]) * 10.0

    import fxu
    import fxu_jacobian
    import hx
    import hx_jacobian
    ekf = ExtendedKalmanFilter(fxu.function, hx.function,
                               fxu_jacobian.function, hx_jacobian.function,
                               Q_ekf, R_ekf, Parameters_ekf, Number_of_Delay)

    # You can create cpp header which can easily define EKF as C++ code
    deployed_file_names = KalmanFilterDeploy.generate_EKF_cpp_code(
        ekf, number_of_delay=Number_of_Delay)
    print(deployed_file_names)

    # %% bicycle model simulation

    sim_delta_time = 0.1
    sim_wheelbase = Parameters_ekf.wheelbase
    simulation_time = 20.0

    def move(x, u, dt, wheelbase):
        hdg = x[2, 0]
        vel = u[0, 0]
        steering_angle = u[1, 0]
        dist = vel * dt

        if abs(steering_angle) > 0.001:  # is robot rotating?
            beta = (dist / wheelbase) * math.tan(steering_angle)
            r = wheelbase / math.tan(steering_angle)  # radius

            dx = np.array([[-r * math.sin(hdg) + r * math.sin(hdg + beta)],
                           [r * math.cos(hdg) - r * math.cos(hdg + beta)],
                           [beta]])
        else:  # moving in straight line
            dx = np.array([[dist * math.cos(hdg)],
                           [dist * math.sin(hdg)],
                           [0]])

        x_next = x + dx + np.array([[np.random.randn() * 0.0],
                                    [np.random.randn() * 0.0],
                                    [np.random.randn() * 0.0]])

        theta = x_next[2, 0]
        # landmark 1
        x_dif = (landmarks[0, 0] - x_next[0, 0])
        y_dif = (landmarks[1, 0] - x_next[1, 0])
        r_1 = math.sqrt(x_dif * x_dif + y_dif * y_dif) + \
            np.random.randn() * 0.0
        phi_1 = math.atan2(y_dif, x_dif) - theta + np.random.randn() * 0.0

        # landmark 2
        x_dif = (landmarks[0, 1] - x_next[0, 0])
        y_dif = (landmarks[1, 1] - x_next[1, 0])
        r_2 = math.sqrt(x_dif * x_dif + y_dif * y_dif) + \
            np.random.randn() * 0.0
        phi_2 = math.atan2(y_dif, x_dif) - theta + np.random.randn() * 0.0

        y_next = np.array([[r_1], [phi_1], [r_2], [phi_2]])

        return x_next, y_next

    def run_simulation():

        x_true = np.array([[2.0], [6.0], [0.3]])  # initial x, y, theta
        u = np.array([[2.0], [0.1]])  # inputs v, steering_angle

        ekf.x_hat = np.array([[0.0], [0.0], [0.0]])  # initial state

        plotter = SimulationPlotter()

        time = np.arange(0, simulation_time, sim_delta_time)

        y_store = [np.zeros((R_ekf.shape[0], 1))] * (Number_of_Delay + 1)
        delay_index = 0

        for i in range(round(simulation_time / sim_delta_time)):
            x_true, y_store[delay_index] = move(x_true, u, sim_delta_time,
                                                sim_wheelbase)  # simulate robot

            # system delay
            delay_index += 1
            if delay_index > Number_of_Delay:
                delay_index = 0

            y_measured = y_store[delay_index]

            # estimate
            ekf.predict(u)
            ekf.update(y_store[delay_index])

            x_estimated = ekf.get_x_hat_without_delay()

            plotter.append(x_true)
            plotter.append(x_estimated)
            plotter.append(y_measured)
            plotter.append(u)

        # plot
        plotter.assign("x_true", column=0, row=0, position=(0, 0),
                       x_sequence=time, label="px_true")
        plotter.assign("x_estimated", column=0, row=0, position=(0, 0),
                       x_sequence=time, label="px_estimated")

        plotter.assign("x_true", column=1, row=0, position=(1, 0),
                       x_sequence=time, label="py_true")
        plotter.assign("x_estimated", column=1, row=0, position=(1, 0),
                       x_sequence=time, label="px_estimated")

        plotter.assign("x_true", column=2, row=0, position=(2, 0),
                       x_sequence=time, label="theta_true")
        plotter.assign("x_estimated", column=2, row=0, position=(2, 0),
                       x_sequence=time, label="theta_estimated")

        plotter.assign("u", column=0, row=0, position=(0, 1),
                       x_sequence=time, label="v")
        plotter.assign("u", column=1, row=0, position=(0, 1),
                       x_sequence=time, label="steering_angle")

        plotter.assign("y_measured", column=0, row=0, position=(1, 1),
                       x_sequence=time, label="r_landmark_1")
        plotter.assign("y_measured", column=2, row=0, position=(1, 1),
                       x_sequence=time, label="r_landmark_2")

        plotter.assign("y_measured", column=1, row=0, position=(2, 1),
                       x_sequence=time, label="angle_landmark_1")
        plotter.assign("y_measured", column=3, row=0, position=(2, 1),
                       x_sequence=time, label="angle_landmark_2")

        plotter.plot("EKF for bicycle model results")

    run_simulation()

    plt.show()


if __name__ == "__main__":
    main()

File: sample/kalman_filter/extended_kalman_filter.cpp
/**
 * @file extended_kalman_filter.cpp
 * @brief Example implementation of an Extended Kalman Filter (EKF) for a
 * bicycle model.
 *
 * This file demonstrates the setup and simulation of an Extended Kalman Filter
 * (EKF) applied to a nonlinear bicycle model. The code defines the plant model,
 * state and measurement functions, their Jacobians, and the EKF structure. It
 * then runs a simulation loop where the true state evolves according to the
 * plant model, measurements are generated, and the EKF estimates the state in
 * the presence of system delay.
 */
#include <iostream>

#include "python_control.hpp"
#include "python_math.hpp"
#include "python_numpy.hpp"

#include <array>

using namespace PythonNumpy;
using namespace PythonControl;

constexpr std::size_t EKF_SIM_STEP_MAX = 200;

constexpr std::size_t NUMBER_OF_DELAY = 5;

constexpr std::size_t STATE_SIZE = 3;
constexpr std::size_t INPUT_SIZE = 2;
constexpr std::size_t OUTPUT_SIZE = 4;

/* model functions */
template <typename T> class BicycleModelParameter {
public:
  BicycleModelParameter(){};

  BicycleModelParameter(const T &delta_time, const T &wheelbase,
                        const T &landmark_1_x, const T &landmark_1_y,
                        const T &landmark_2_x, const T &landmark_2_y)
      : delta_time(delta_time), wheelbase(wheelbase),
        landmark_1_x(landmark_1_x), landmark_1_y(landmark_1_y),
        landmark_2_x(landmark_2_x), landmark_2_y(landmark_2_y) {}

public:
  T delta_time;
  T wheelbase;
  T landmark_1_x;
  T landmark_1_y;
  T landmark_2_x;
  T landmark_2_y;
};

template <typename T>
auto bicycle_model_state_function(const StateSpaceState_Type<T, STATE_SIZE> &X,
                                  const StateSpaceInput_Type<T, INPUT_SIZE> &U,
                                  const BicycleModelParameter<T> &parameters)
    -> StateSpaceState_Type<T, STATE_SIZE>;

template <typename T, typename A_Type>
auto bicycle_model_state_function_jacobian(
    const StateSpaceState_Type<T, STATE_SIZE> &X,
    const StateSpaceInput_Type<T, INPUT_SIZE> &U,
    const BicycleModelParameter<T> &parameters) -> A_Type;

template <typename T>
auto bicycle_model_measurement_function(
    const StateSpaceState_Type<T, STATE_SIZE> &X,
    const BicycleModelParameter<T> &parameters)
    -> StateSpaceOutput_Type<T, OUTPUT_SIZE>;

template <typename T, typename C_Type>
auto bicycle_model_measurement_function_jacobian(
    const StateSpaceState_Type<T, STATE_SIZE> &X,
    const BicycleModelParameter<T> &parameters) -> C_Type;

int main(void) {
  /* Create plant model */
  using X_Type = StateSpaceState_Type<double, STATE_SIZE>;
  using U_Type = StateSpaceInput_Type<double, INPUT_SIZE>;
  using Y_Type = StateSpaceOutput_Type<double, OUTPUT_SIZE>;

  using SparseAvailable_A =
      SparseAvailable<ColumnAvailable<true, false, true>,
                      ColumnAvailable<false, true, true>,
                      ColumnAvailable<false, false, true>>;

  using A_Type = SparseMatrix_Type<double, SparseAvailable_A>;

  using SparseAvailable_C = SparseAvailable<
      ColumnAvailable<true, true, false>, ColumnAvailable<true, true, true>,
      ColumnAvailable<true, true, false>, ColumnAvailable<true, true, true>>;

  using C_Type = SparseMatrix_Type<double, SparseAvailable_C>;

  auto Q = make_KalmanFilter_Q<STATE_SIZE>(1.0, 1.0, 1.0);

  using Q_Type = decltype(Q);

  auto R = make_KalmanFilter_R<OUTPUT_SIZE>(10.0, 10.0, 10.0, 10.0);

  using R_Type = decltype(R);

  /* Parameters */
  using Parameter_Type = BicycleModelParameter<double>;

  Parameter_Type parameters(0.1, 0.5, -1.0, -1.0, 10.0, 10.0);

  /* state and measurement functions */
  StateFunction_Object<X_Type, U_Type, BicycleModelParameter<double>>
      state_function = bicycle_model_state_function<double>;

  StateFunctionJacobian_Object<A_Type, X_Type, U_Type,
                               BicycleModelParameter<double>>
      state_function_jacobian =
          bicycle_model_state_function_jacobian<double, A_Type>;

  MeasurementFunction_Object<Y_Type, X_Type, BicycleModelParameter<double>>
      measurement_function = bicycle_model_measurement_function<double>;

  MeasurementFunctionJacobian_Object<C_Type, X_Type,
                                     BicycleModelParameter<double>>
      measurement_function_jacobian =
          bicycle_model_measurement_function_jacobian<double, C_Type>;

  /* define EKF */
  ExtendedKalmanFilter<A_Type, C_Type, U_Type, Q_Type, R_Type, Parameter_Type,
                       NUMBER_OF_DELAY>
      ekf(Q, R, state_function, state_function_jacobian, measurement_function,
          measurement_function_jacobian, parameters);

  /* simulation */
  auto x_true_initial = make_StateSpaceState<STATE_SIZE>(2.0, 6.0, 0.3);
  decltype(x_true_initial) x_true;

  auto u = make_StateSpaceInput<INPUT_SIZE>(2.0, 0.1);

  ekf.X_hat.template set<0, 0>(0.0);
  ekf.X_hat.template set<1, 0>(0.0);
  ekf.X_hat.template set<2, 0>(0.0);

  std::array<StateSpaceOutput_Type<double, OUTPUT_SIZE>, (NUMBER_OF_DELAY + 1)>
      y_store;

  std::size_t delay_index = 0;

  x_true = x_true_initial;
  for (std::size_t i = 0; i < EKF_SIM_STEP_MAX; i++) {
    x_true = bicycle_model_state_function<double>(x_true, u, parameters);
    y_store[delay_index] =
        bicycle_model_measurement_function<double>(x_true, parameters);

    // system delay
    delay_index++;
    if (delay_index > NUMBER_OF_DELAY) {
      delay_index = 0;
    }

    ekf.predict(u);
    ekf.update(y_store[delay_index]);

    for (std::size_t j = 0; j < STATE_SIZE; j++) {
      std::cout << "X_hat[" << j << "]: " << ekf.get_x_hat_without_delay()(j, 0)
                << ", ";
    }
    std::cout << std::endl;
  }
  std::cout << std::endl;

  for (std::size_t j = 0; j < STATE_SIZE; j++) {
    std::cout << "x_true[" << j << "]: " << x_true(j, 0) << ", ";
  }
  std::cout << std::endl;

  return 0;
}

template <typename T>
auto bicycle_model_state_function(const StateSpaceState_Type<T, STATE_SIZE> &X,
                                  const StateSpaceInput_Type<T, INPUT_SIZE> &U,
                                  const BicycleModelParameter<T> &parameters)
    -> StateSpaceState_Type<T, STATE_SIZE> {

  using namespace PythonMath;

  T x = X.template get<0, 0>();
  T y = X.template get<1, 0>();
  T theta = X.template get<2, 0>();
  T v = U.template get<0, 0>();
  T steering_angle = U.template get<1, 0>();

  T wheelbase = parameters.wheelbase;
  T delta_time = parameters.delta_time;

  return StateSpaceState_Type<T, STATE_SIZE>(
      {{-wheelbase * sin(theta) / tan(steering_angle) +
        wheelbase *
            sin(delta_time * v * tan(steering_angle) / wheelbase + theta) /
            tan(steering_angle) +
        x},
       {wheelbase * cos(theta) / tan(steering_angle) -
        wheelbase *
            cos(delta_time * v * tan(steering_angle) / wheelbase + theta) /
            tan(steering_angle) +
        y},
       {delta_time * v * tan(steering_angle) / wheelbase + theta}});
}

template <typename T, typename A_Type>
auto bicycle_model_state_function_jacobian(
    const StateSpaceState_Type<T, STATE_SIZE> &X,
    const StateSpaceInput_Type<T, INPUT_SIZE> &U,
    const BicycleModelParameter<T> &parameters) -> A_Type {

  using namespace PythonMath;

  T theta = X.template get<2, 0>();
  T v = U.template get<0, 0>();
  T steering_angle = U.template get<1, 0>();

  T wheelbase = parameters.wheelbase;
  T delta_time = parameters.delta_time;

  A_Type A;

  A.template set<0, 0>(1);
  A.template set<0, 1>(0);
  A.template set<0, 2>(
      -wheelbase * cos(theta) / tan(steering_angle) +
      wheelbase *
          cos(delta_time * v * tan(steering_angle) / wheelbase + theta) /
          tan(steering_angle));
  A.template set<1, 0>(0);
  A.template set<1, 1>(1);
  A.template set<1, 2>(
      -wheelbase * sin(theta) / tan(steering_angle) +
      wheelbase *
          sin(delta_time * v * tan(steering_angle) / wheelbase + theta) /
          tan(steering_angle));
  A.template set<2, 0>(0);
  A.template set<2, 1>(0);
  A.template set<2, 2>(1);

  return A;
}

template <typename T>
auto bicycle_model_measurement_function(
    const StateSpaceState_Type<T, STATE_SIZE> &X,
    const BicycleModelParameter<T> &parameters)
    -> StateSpaceOutput_Type<T, OUTPUT_SIZE> {

  using namespace PythonMath;

  T x = X.template get<0, 0>();
  T y = X.template get<1, 0>();
  T theta = X.template get<2, 0>();

  T landmark_1_x = parameters.landmark_1_x;
  T landmark_2_x = parameters.landmark_2_x;
  T landmark_2_y = parameters.landmark_2_y;
  T landmark_1_y = parameters.landmark_1_y;

  T dif_1_x = landmark_1_x - x;
  T dif_1_y = landmark_1_y - y;
  T dif_2_x = landmark_2_x - x;
  T dif_2_y = landmark_2_y - y;

  return StateSpaceOutput_Type<T, OUTPUT_SIZE>(
      {{sqrt(dif_1_x * dif_1_x + dif_1_y * dif_1_y)},
       {-theta + atan2(dif_1_y, dif_1_x)},
       {sqrt(dif_2_x * dif_2_x + dif_2_y * dif_2_y)},
       {-theta + atan2(dif_2_y, dif_2_x)}});
}

template <typename T, typename C_Type>
auto bicycle_model_measurement_function_jacobian(
    const StateSpaceState_Type<T, STATE_SIZE> &X,
    const BicycleModelParameter<T> &parameters) -> C_Type {

  using namespace PythonMath;

  T x = X.template get<0, 0>();
  T y = X.template get<1, 0>();

  T landmark_1_x = parameters.landmark_1_x;
  T landmark_2_x = parameters.landmark_2_x;
  T landmark_2_y = parameters.landmark_2_y;
  T landmark_1_y = parameters.landmark_1_y;

  C_Type C;

  T dif_1_x = landmark_1_x - x;
  T dif_1_y = landmark_1_y - y;
  T dif_2_x = landmark_2_x - x;
  T dif_2_y = landmark_2_y - y;

  C.template set<0, 0>((-landmark_1_x + x) /
                       sqrt(dif_1_x * dif_1_x + dif_1_y * dif_1_y));
  C.template set<0, 1>((-landmark_1_y + y) /
                       sqrt(dif_1_x * dif_1_x + dif_1_y * dif_1_y));
  C.template set<0, 2>(0);
  C.template set<1, 0>(-(-landmark_1_y + y) /
                       (dif_1_x * dif_1_x + dif_1_y * dif_1_y));
  C.template set<1, 1>(-(landmark_1_x - x) /
                       (dif_1_x * dif_1_x + dif_1_y * dif_1_y));
  C.template set<1, 2>(-1);
  C.template set<2, 0>((-landmark_2_x + x) /
                       sqrt(dif_2_x * dif_2_x + dif_2_y * dif_2_y));
  C.template set<2, 1>((-landmark_2_y + y) /
                       sqrt(dif_2_x * dif_2_x + dif_2_y * dif_2_y));
  C.template set<2, 2>(0);
  C.template set<3, 0>(-(-landmark_2_y + y) /
                       (dif_2_x * dif_2_x + dif_2_y * dif_2_y));
  C.template set<3, 1>(-(landmark_2_x - x) /
                       (dif_2_x * dif_2_x + dif_2_y * dif_2_y));
  C.template set<3, 2>(-1);

  return C;
}
